chronicles/textlineserializer.nim

```diff
@@ -26,47 +28,63 @@ serializationFormat TextLineLog,
                     mimeType = "text/plain"
 
 #
-# color support functions
+# value support functions
 #
-proc fgColor(writer: TextLineWriter, color: ForegroundColor, brightness: bool) =
-  when writer.colorScheme == AnsiColors:
-    writer.stream.writeText ansiForegroundColorCode(color, brightness)
-  when writer.colorScheme == NativeColors:
-    writer.stream.setForegroundColor(color, brightness)
-
-proc resetColors(writer: TextLineWriter) =
-  when writer.colorScheme == AnsiColors:
-    writer.stream.writeText ansiResetCode
-  when writer.colorScheme == NativeColors:
-    writer.stream.resetAttributes()
-
-proc applyStyle(writer: TextLineWriter, style: Style) =
-  when writer.colorScheme == AnsiColors:
-    writer.stream.writeText ansiStyleCode(style)
-  when writer.colorScheme == NativeColors:
-    writer.stream.setStyle({style})
+
+const
+  escChars: set[char] = strutils.NewLines + {'"', '\\'}
+  quoteChars: set[char] = {' ', '='}
+
+type
+  typesNotNeedingQuotes = float | float64 | float32 | uint8 | uint16 | uint32 | uint64 | int | int8 | int16 | int32 | int64
+
+proc quoteIfNeeded(w: var TextLineWriter, val: typesNotNeedingQuotes) =
+  w.stream.writeText val
+
+proc quoteIfNeeded(w: var TextLineWriter, val: auto) =
+  let valText = $val
+  let
+    needsEscape = valText.find(escChars) > -1
+    needsQuote = (valText.find(quoteChars) > -1) or needsEscape
+  if needsQuote:
+    var quoted = ""
+    quoted.addQuoted valText
+    w.stream.writeText quoted
+  else:
+    w.stream.writeText val
+
+proc quoteIfNeeded(w: var TextLineWriter, val: ref Exception) =
+  w.stream.writeText val.name
+  w.stream.writeText '('
+  w.quoteIfNeeded val.msg
+  when not defined(js) and not defined(nimscript) and hostOS != "standalone":
+    w.stream.writeText ", "
+    w.quoteIfNeeded getStackTrace(val).strip
+  w.stream.writeText ')'
 
 #
 # Class startup
 #
+
 proc init*(w: var TextLineWriter, stream: OutputStream) =
   w.stream = stream
 
 #
 # Field Handling
 #
+
 proc writeFieldName*(w: var TextLineWriter, name: string) =
   w.stream.writeText ' '
   let (color, bright) = levelToStyle(w.currentLevel)
-  w.fgColor(color, bright)
+  w.setForegroundColor(color, bright)
   w.stream.writeText name
-  w.resetColors()
+  w.resetAllColors()
   w.stream.writeText "="
 
 proc writeValue*(w: var TextLineWriter, value: auto) =
-  w.fgColor(propColor, true)
-  w.stream.writeTextQuoted(value, optional=true)
-  w.resetColors()
+  w.setForegroundColor(propColor, true)
```

Comment:
This may look like a strange request, but the dot operator generally doesn't work reliably in generic functions (in some complex instantiation scenarios, you'll get an error `undefined identifier: setForegroundColor`. That's why most of the code in Chronicles was using the more C-like `setForegroundColor(w, propColor, true)` syntax.

It's not a personal preference of mine, but rather something I've discovered the hard way.

Reply:
changed.